### PR TITLE
tests: add unit tests for convert_parse_tlvs()

### PR DESCRIPTION
**Pull request by GregoryVds**
Signed-off-by: Gregory Vander Schueren <gregory.vanderschueren@tessares.net>
Change-Id: I457daa11f299590544cd7ae9c28b3eae31a39d71

### CODE REVIEW

File: tests/check_convert_util.c [hunk 1]
```diff
@@ -64,9 +96,116 @@ START_TEST(test_convert_parse_header){
 }
 END_TEST
 
-START_TEST(test_convert_parse_tlvs){
-	/* TODO */
-	ck_assert(1);
+START_TEST(test_convert_parse_tlvs_generic){
```
Comment (matttbe) [inline tests/check_convert_util.c:99]: detail: you might want to add `macro-open START_TEST` in your uncrustify config (and remove all the other specific names from the project you stole this config ;-) )

File: tests/check_convert_util.c [hunk 2]
```diff
@@ -64,9 +96,116 @@ START_TEST(test_convert_parse_header){
...
+	tlv	= (struct convert_tlv *)buff;
+
+	ret = convert_parse_tlvs(buff, 0, &opts);
+	ck_assert_msg(ret == -1, "Should fail: 0-len buff");
+
+	ret = convert_parse_tlvs(buff, sizeof(struct convert_tlv) - 1, &opts);
+	ck_assert_msg(ret == -1,
+	              "Should fail: buff len shorter than TLV header");
+
+	ret = convert_parse_tlvs(buff, buff_len - 1, &opts);
```
Comment (matttbe) [inline tests/check_convert_util.c:116]: (this is more to test your own test functions but why not)

File: tests/check_convert_util.c [hunk 3]
```diff
@@ -64,9 +96,116 @@ START_TEST(test_convert_parse_header){
...
+	ck_assert_msg(ret == -1,
+	              "Should fail: buff len shorter than TLV length");
+
+	tlv->type	= 42;
+	ret		= convert_parse_tlvs(buff, buff_len, &opts);
```
Comment (matttbe) [inline tests/check_convert_util.c:121]: should you also try with something which does not fail? :)

**Review by matttbe — COMMENTED** [review]
just two questions, nothing important ;-)

File: tests/check_convert_util.c (hunk 2 above)
Comment (GregoryVds) [reply tests/check_convert_util.c:116]: This indeed assumes that sample_convert_connect() is correct :-) But assuming sample_convert_connect() is correct, this validates that convert_parse_tlvs() correctly detects a truncated packet. I could move the logic from sample_convert_error() to here, but the problem would remain the same in the end, the testing code has to be correct.

File: tests/check_convert_util.c (hunk 3 above)
Comment (GregoryVds) [reply tests/check_convert_util.c:121]: The success case is handled in the dedicated function for each type of TLV.
Adding a success case here would imply duplicating a test.

File: tests/check_convert_util.c (hunk 1 above)
Comment (GregoryVds) [reply tests/check_convert_util.c:99]: Thanks for the macro-open tip, will add it. I thought I removed everything specific to the other project, but indeed I forgot some stuff at the end, will remove that :-)

**Review by GregoryVds — COMMENTED** [review]
Thanks for the review Matt! Waiting for you answer to my answers :-) I think I can leave the code like this, I will fix the Uncrustify config in another patch.

**Review by matttbe — APPROVED** [review]

**Pull request merged**